Chapters 1 & 2 Revisions 05-02-2025 (#5113)
dfinity/motoko: doc/md/fundamentals/2-types/5-records.md

> # Records

Author: i feel like some don't need to be runnable but at this point i think its all runnable or not runnable at all, mixture makes it slightly confusing at least per page consistency would be good.

# Records

> Records let you group related values using named fields. Each field can have a different type. Unlike [tuples](https://internetcomputer.org/docs/motoko/fundamentals/types/tuples), which rely on positional access, records use field names for retrieval, improving maintainability. Records also support mutable fields, declared using the `var` keyword.  The fields of a tuple are always immutable.

Reviewer: ```suggestion
Records allow you to group related values using named fields, with each field potentially having a different type.  
Unlike [tuples](https://internetcomputer.org/docs/motoko/fundamentals/types/tuples), which use positional access, records provide field-based access, improving readability and maintainability.

Records also support **mutable fields**, declared using the `var` keyword.  
In contrast, all fields in a tuple are always **immutable**.
```

};
```

:::info Inferred types

> Type annotations on immutable record fields aren’t necessary as the compiler will infer them.

Reviewer: ```suggestion
Type annotations on immutable record fields are optional, as the compiler can infer their types automatically.
```

> However, for mutable fields (`var`), it’s good practice to explicitly declare the type. Otherwise, the compiler might infer a more specific type than intended, preventing future assignments of a broader type.

Reviewer: ```suggestion
However, for mutable fields (`var`), it's good practice to explicitly declare the type.  
Without an explicit annotation, the compiler might infer a more specific type than intended, which can restrict future assignments to broader types.
```

Example:

```motoko
let account = { name = "Motoko"; var balance = 0 };
// Inferred as { name : Text; var balance : Nat }
account.balance := -100; //  Rejected, -100 is not a Nat
````

To avoid this issue, annotate the field explicitly:

```motoko
let account = { name = "Motoko"; var balance : Int = 0 };
account.balance := -100; //  Allowed
```

> This recommendation also applies to all `var` declarations—not just record fields.

Reviewer: ```suggestion
This recommendation also applies to all `var` declarations, not just record fields.
```

};
```

:::info Inferred types
Type annotations on immutable record fields are optional, as the compiler can infer their types automatically.
However, for mutable fields (`var`), it's good practice to explicitly declare the type.  
Without an explicit annotation, the compiler might infer a more specific type than intended, which can restrict future assignments to broader types.

Example:

```motoko
let account = { name = "Motoko"; var balance = 0 };
// Inferred as { name : Text; var balance : Nat }
account.balance := -100; //  Rejected, -100 is not a Nat
````

To avoid this issue, annotate the field explicitly:

> ```motoko

Reviewer: does not return meaningful output

> ```motoko

Reviewer: This did not get applied.

person.name := "Ghost";  // Now person.name is "Ghost"
person.age := 30;      // Now person.age is 30

person
```

Attempting to update an immutable field is a compile-time type error.

## Nested records

> 

Author: This does not need an actor it's a simple assignment, its also no-repl.

Records can be destructured using [`switch`](https://internetcomputer.org/docs/motoko/fundamentals/control-flow/switch), allowing selective extraction of fields. This approach makes accessing deeply nested fields more explicit and readable.

> 

Reviewer 2: The intro on line 7 could be improved (unfortunately, I can't suggest it).
Old:
```
Records provide a structured way to group related values using named fields. Unlike [tuples](https://internetcomputer.org/docs/motoko/fundamentals/types/tuples), which rely on positional access, records use field names for retrieval, improving clarity and maintainability. Unlike tuples, record also support mutable fields.  
```
new:
```
Records let you group related values using named fields. Each field can have a different type. Unlike [tuples](https://internetcomputer.org/docs/motoko/fundamentals/types/tuples), which rely on positional access, records use field names for retrieval, improving maintainability. Records also support mutable fields, declared using the `var` keyword.  The fields of a tuple are always immutable.
``

Reviewer 2: Can you make the example runnable. Also, some of the examples end in Debug.prints, which require you to also import Debug. Instead, You can remove the debug print and the repl will just show the value of the last declaration:

For example:
```
import Debug "mo:base/Debug":
let person = {
   var name : Text = "Motoko";
   var age : Nat = 25;
  };

person.name := "Ghost";  // Now person.name is "Ghost"
person.age := 30;      // Now person.age is 30

Debug.print(person.name # is # debug_show(person.age));
```
Could just be:
```
import Debug "mo:base/Debug":
let person = {
   var name : Text = "Motoko";
   var age : Nat = 25;
  };

person.name := "Ghost";  // Now person.name is "Ghost"
person.age := 30;      // Now person.age is 30

person;
```
When run, that will display the current value of person:
```
{age = 25; name = "Ghost"} : {var age : Nat; var name : Text}
```

> 

Reviewer 2: Above: replace:
```
Attempting to update an immutable field will raise an error.
```
by
```
Attempting to update an immutable field is a compile-time type error.
```
To make clear this isn't a runtime error.

Author: This is a pretty good distinction ! i'm not too sure where to put it because it might be more relevant in the records section because the example uses a record so i will leave an admonition for now and we can make a final decision later.

> ```motoko no-repl

Reviewer 2: ```suggestion
```motoko
```

| Pattern matching| Complete, using ordered tuple patterns                       | Selective, using unordered record patterns  |  
| Mutability     | Immutable after creation                 | Can have mutable fields               |
| Naming         | Fields are anonymous  | Fields are named               |  
| Subtyping     | Fields cannot be removed | Fields can be removed |

> | Use Case       | Positional grouping of related values, e.g. vectors     |  structured data types        |

Reviewer: ```suggestion
| Use case       | Positional grouping of related values, e.g. vectors     |  Structured data types        |
```

Motoko's records afford more flexible subtyping than tuples. Subtyping on records allows fields to be removed in the subtype, while for tuples, subtyping always requires the length of the tuples to be equal.

> For example,, `{x : Int, y : Int, z : Int}` is a subtype of `{x : Int, y : Int}`, but `(Int, Int, Int)` is not a subtype of `(Int, Int)`.  

Reviewer 2: ```suggestion
For example, `{x : Int, y : Int, z : Int}` is a subtype of `{x : Int, y : Int}`, but `(Int, Int, Int)` is not a subtype of `(Int, Int)`.  
```

| Pattern matching| Complete, using ordered tuple patterns                       | Selective, using unordered record patterns  |  
| Mutability     | Immutable after creation                 | Can have mutable fields               |
| Naming         | Fields are anonymous  | Fields are named               |  
| Subtyping     | Fields cannot be removed | Fields can be removed |
| Use case       | Positional grouping of related values, e.g. vectors     |  Structured data types        |

> Motoko's records afford more flexible subtyping than tuples. Subtyping on records allows fields to be removed in the subtype, while for tuples, subtyping always requires the length of the tuples to be equal.

Reviewer: ```suggestion
Motoko's records support more flexible subtyping than tuples.  
With records, subtyping allows fields to be omitted in the subtype (a concept known as **width subtyping**).  
In contrast, tuple subtyping requires tuples to have the **same length**, making them less flexible in this regard.
```
